DNS Resolver

mhrabovcin opened the pull request:
Update static `leader.mesos` name in nginx configuration to variables so it will get re-resolved based on record TTL.

mhrabovcin committed:
resolver: make sure that leader.mesos is resolved every 30 seconds

mhrabovcin committed:
add nginx resolver / mesos leader test

jgehrcke reviewed (changes requested):
Hey Martin, I quickly went through that and left a few comments. 

If everyone agrees that we go ahead with this -- then we must definitely add an expressive comment or a README section about this DNS re-resolution trick through variable usage.

alberts commented on nginx.master.conf:
```diff
@@ -4,26 +4,18 @@ include common/main.conf;
 
     upstream mesos_dns {
-        server master.mesos:8123;
+        server 127.0.0.1:8123;
```
Agreed. I think we want this change, but let's separate any changes from master.mesos to 127.0.0.1 into a separate PR

mhrabovcin replied:
I did this to be consistent with upstream - https://github.com/mesosphere/adminrouter/blob/enterprise/nginx.master.conf#L17-L19

jgehrcke replied:
OK, then I am good with having it in this PR. 

@alberts @vespian please confirm.

mhrabovcin commented on nginx.master.conf:
```diff
@@ -4,26 +4,18 @@ include common/main.conf;
     # Without this, cosocket-based code in worker
     # initialization cannot resolve leader.mesos.
-    resolver 127.0.0.1:61053;
+    resolver 127.0.0.1:61053 ipv6=off;
```
We're not doing IPv6 so this will save us some queries to DNS backend that would get empty replies anyways.

jgehrcke replied:
Although that's a change in behavior, we're probably good with it :-).

mhrabovcin commented on nginx.master.conf:
```diff
@@ -4,26 +4,18 @@ include common/main.conf;
-
-    upstream dcos_history_service {
-        server leader.mesos:15055;
+        server 127.0.0.1:8080;
```
Yes and that is still true but done through variable now: https://github.com/dcos/adminrouter/pull/50/files#diff-8a636580686895dc929d495101122338R247

jgehrcke replied:
But then we can remove this upstream definition, can't we?

mhrabovcin replied:
I guess that diff is borked for you somehow because this PR is removing the `dcos_history_service` upstream definition.

jgehrcke replied:
Turns out I am borked. My brain reproducibly ignored the fact that the `-    upstream dcos_history_service` line has minus and is read. I only consciously saw the change from leader.mesos to 127.0.0.1. Sorry!

jgehrcke reviewed (approved):
Great work!

vespian commented on test-harness/tests/conftest.py:
```diff
@@ -95,32 +95,16 @@ def syslog_mock(log_catcher):
-
-
-@pytest.fixture(scope='session')
-def navstar_ips():
```
Why do we remove these? Please remember that agent uses them: https://github.com/dcos/adminrouter/blob/master/nginx.agent.conf#L5

mhrabovcin replied:
I've added back `navstar_ips` fixture and moved the dependency to the `nginx_class` fixture. One thing I don't understand though is what is being resolved with these IP addresses in case of `agent` AR instance. I'd expect some tests to fail if these were required.

vespian replied:
Please remember that we have not written all the tests yet. We have ~130, I stimate we need 600-700 to have full coverage. IIRC Agent Adminrouter is almost not tested yet.

jgehrcke replied:
> we have not written all the tests yet

I understand that the tests you are referring to here are supposed to cover basic behavior.

Just noting: in case of Admin Router, "full coverage" is undefined if we talk about real-world usage scenarios. "full coverage" in terms of code coverage is great, but can easily lead to a wrong sense of confidence. Concepts, formalism plus fuzzing/chaos/statistical testing with a high sampling density are required to obtain the degree of confidence that we strive for.

vespian replied:
http://stackoverflow.com/a/90021

vespian commented on test-harness/tests/conftest.py:
```diff
@@ -7,10 +7,10 @@
 import pytest
 
 from runner.common import (
-    DNSMock,
```
Could you also please clean up the DNSMock class and all stuff related to dnsmasq from `docker/` container?

vespian commented on test-harness/modules/mocker/dns.py:
```diff
@@ -0,0 +1,93 @@
+    def stop(self):
+        self.server.stop()
+
+    def reply_with_leader_ip(self, leader_ip):
```
It would be nice to manipulate domain TTL in the same manner, if this is required by a test. Then we can make default very high, and use low values for tests that require it. We do not have to worry about AR caching it, as we controll AR and can simply launc our own instance the same way it's done in `test_cache.py`.

vespian commented on test-harness/tests/test_master.py:
```diff
@@ -204,3 +205,36 @@ def test_if_upstream_headers_are_correct(self,
+
+        # Revert back to the original Mesos leader
+        dns_server_mock.reply_with_leader_ip('127.0.0.2')
+        sleep(2)
```
Please have a look at `test_cache.py` and how a new AR instance is started for each tests. Using this approach you will not have to introduce artificial delays/slow down the tests.

vespian commented on test-harness/tests/test_master.py:
```diff
@@ -204,3 +205,36 @@ def test_if_upstream_headers_are_correct(self,
+        assert r.json()["endpoint_id"] == "http://127.0.0.5:5050"
+
+        # Revert back to the original Mesos leader
+        dns_server_mock.reply_with_leader_ip('127.0.0.2')
```
This should be done by a fixture cleanup, have a look i.e. at: https://github.com/dcos/adminrouter/blob/master/test-harness/tests/conftest.py#L63-L74

vespian commented on test-harness/tests/test_master.py:
```diff
@@ -204,3 +205,36 @@ def test_if_upstream_headers_are_correct(self,
+        assert r.json()["cluster"] == "prozlach-qzpz04t"
+
+        # See MockerBase::_create_common_endpoints() where we create ReflectingEndpoint
+        dns_server_mock.reply_with_leader_ip('127.0.0.5')
```
We really need this test to be launched in a separate AR instance that is GCed after the tests finishes.

If we do it using shared AR instance (long story short - all that is in the `tests_master.py` file, more info in the README.md) we are risking creating HeisenBugs/flaky tests as change of leader.mesos may be cached i.e. by lua code.

vespian commented on test-harness/tests/test_master.py:
```diff
@@ -204,3 +205,36 @@ def test_if_upstream_headers_are_correct(self,
+        # instance of ReflectingTcpIpEndpoint
+        r = requests.get(url, allow_redirects=True, headers=valid_user_header)
+        assert r.status_code == 200
+        assert r.json()["endpoint_id"] == "http://127.0.0.5:5050"
```
What about testing Lua code? Do we need to do it?

https://github.com/dcos/adminrouter/blob/master/master/cache.lua#L263

vespian commented on test-harness/tests/test_master.py:
```diff
@@ -204,3 +205,36 @@ def test_if_upstream_headers_are_correct(self,
+        assert r.status_code == 200
+        assert r.json()["cluster"] == "prozlach-qzpz04t"
+
+        # See MockerBase::_create_common_endpoints() where we create ReflectingEndpoint
```
Let's skip this. Currently there are ~15 different endpoints, we will not be able to comment all tests with "which endpoints is created where" type of comments :)

vespian reviewed (changes requested):
Left some comments.

The DNS server you wrote is briliant. No matter the fate of this PR, it would be great if it was merged into the test-harness. I only wish that I knew about it earlier :)

mhrabovcin committed:
PR feedback

mhrabovcin commented on test-harness/modules/mocker/dns.py:
```diff
@@ -0,0 +1,93 @@
+
+    def resolve(self, request, handler):
+        # We can respond only to leader.mesos A requests
+        if (str(request.q.qname) == (self.domain.leader + ".") and
```
I've removed remaining entries because they weren't being used in any test. Is there some other reason to include these?

vespian replied:
They are not used at the moment - true.

But the thing is that we want other people to write tests as well - let's make it easier for them. It costs us very little to do it, and it may save them some time if/when they need it.

vespian replied:
https://github.com/mesosphere/dcos-enterprise/blob/master/gen_extra/dcos-config.yaml#L88

AR agent uses `master.mesos`. Let's provide all the aliases please.

mhrabovcin commented:
I've addressed most of the feedback in a separate commit so its easier to review changes.

* Move new tests to a separate file 
* Run AR process per test (similar to pattern in `test_cache.py`)
* Added extra tests that are exercising DNS resolution of lua code
* Improved documentation about DNS resolution
* Cleaned up `dnsmask` dependencies
* Improved clarity of DNS mock library (TTL)

vespian commented on README.md:
```diff
@@ -32,6 +32,34 @@ The endpoint should only use relative links for links and referenced assets such
 
 Tasks running in nested [Marathon app groups](https://mesosphere.github.io/marathon/docs/application-groups.html) will be available only using their service name (i.e, `<dcos-cluster>/service/<service-name>`) and not considering the marathon app group name (i.e., `<dcos-cluster>/service/app-group/<service-name>`).
 
+## DNS resolution
```
❤️ 

Thanks for writting the docs!

vespian commented on README.md:
```diff
@@ -166,13 +194,19 @@ and `.stop()` methods during the start and stop of mocker instance
+#### DNS mock
+
+The AR requires a working DNS server that responds correctly to
+`leader.mesos` - `A` type query and resolves current Mesos leader instance.
```
We will have to update it, see my earlier comments.

vespian commented on test-harness/tests/conftest.py:
```diff
@@ -95,23 +95,37 @@ def syslog_mock(log_catcher):
-    m_61053.stop()
-    m_53.stop()
+
+@pytest.fixture(scope='function')
```
+1

vespian commented on test-harness/tests/test_resolver.py:
```diff
@@ -0,0 +1,121 @@
+            # its enough to wait for 2 seconds here
+            time.sleep(self.SHORT_TTL * 2)
+
+            # Make sure that second request was forwarded to new leader
```
`to the new`

vespian commented on test-harness/tests/test_resolver.py:
```diff
@@ -0,0 +1,121 @@
+            assert r.status_code == 200
+            assert r.json()["endpoint_id"] == "http://127.0.0.5:5050"
+
+    def test_if_mesos_leader_ar_instance_is_reresolved_in_proxy_pass(
```
What is the difference between this test and the previous one? Maybe it's too early in the morning for me :(

mhrabovcin replied:
Its testing different configuration paths, but same resolver functionality. One difference here is that previous test requests end at Mesos servers running on port `5050` while this one tests redirecting to other AR instance running on current Mesos leader on port `80`. 

I am happy to remove this test.

vespian replied:
I see now - no, we should definitely have it - I was missing something like this when I was writing leader endpoints tests.

I would like to suggest a different approach though - instead of copy-pasting code, we can:
* add two new reflecting endpoints which will simulate extra Mesos Master processes, make them have endpoint IDs A and B.
* parametrize the test with path(`/system/v1/leader/mesos/foo/bar`, `/mesos/master/state-summary`)
* make your DNS mock server resolve `leader.mesos` as address A
* launch AR in GuardedSubprocess context
* verify that the call to path returns endpoint A
* make your DNS mock server resolve `leader.mesos` as address B
* verify that the call to path returns endpoint A

AR will not like the fact that leader.mesos is a reflector and will bork, but this is irrelevant to our test.

The ideal solution for me would be to make Mesos/Marathon endpoints have a reflecting API location, but this is nontrivial and outside of the scope of this PR.

vespian commented on test-harness/tests/test_resolver.py:
```diff
@@ -0,0 +1,121 @@
+
+        with GuardedSubprocess(ar):
+            # Let the cache warm-up:
+            time.sleep(CACHE_FIRST_POLL_DELAY + 1)
```
You can also issue a request (i.e. call `ping_mesos_agent`) and trigger the refresh earlier so there is no need make the tests slower than we need.

vespian commented on test-harness/tests/test_resolver.py:
```diff
@@ -0,0 +1,121 @@
+        with GuardedSubprocess(ar):
+            # Let the cache warm-up:
+            time.sleep(CACHE_FIRST_POLL_DELAY + 1)
+            for _ in range(3):
```
Why three times?

vespian reviewed (commented):
Some more comments for you.

I think you did a great job at testing this change outside-in. It seems I was wrong and we do can reliably test DNS re-resolving :)

vespian reviewed (changes requested):
Left a reply to your comment. Please check my earlier comments too. Thanks!

mhrabovcin committed:
make DNS server to respond to all necessary queries

mhrabovcin commented:
Next iteration:

* Simplified leader resolution in `proxy_pass` directive test
* Updated DNS server to include all required `*.mesos` domains

vespian commented on test-harness/modules/mocker/common.py:
```diff
@@ -53,6 +53,13 @@ def _create_common_endpoints():
+        res.append(MesosEndpoint(ip='127.0.0.3', port=5050))
+        # mesos non-leading master that runs Mesos and AR
+        res.append(ReflectingTcpIpEndpoint(ip='127.0.0.10', port=5050))
+        res.append(ReflectingTcpIpEndpoint(ip='127.0.0.10', port=80))
```
No need for these, we already have `ip='127.0.0.2', port=80` and `ip='127.0.0.3', port=80` a couple of lines below. Every time you add a new Endpoint it makes tests start&stop a bit slower so let's keep this number at minimum.

mhrabovcin replied:
I understand but Mesos mocked servers running on port `5050` on both IP addresses aren't `ReflectingTcpIpEndpoint` but rather `MesosEndpoint`. This means that we couldn't use parametrized test and single `generic_correct_upstream_dest_test` for both type of URLs.

vespian replied:
Then let's remove the other two - they are used for marathon leader testing if I remember correctly.

Adjusting the tests should not be hard.

mhrabovcin replied:
Quite a lot of tests are failing:

```
38 failed, 95 passed in 50.86 seconds
```

It looks like most of cache tests expect `127.0.0.2` to be returning actual mesos state response.

vespian replied:
k, will have a look at it later.

vespian commented on test-harness/modules/mocker/dns.py:
```diff
@@ -0,0 +1,144 @@
+        rclass=1,
+        ttl=DEFAULT_TTL,
+        rdata=A('127.0.0.1')),
+}
```
👍

vespian commented on test-harness/tests/test_resolver.py:
```diff
@@ -0,0 +1,100 @@
+    @pytest.mark.parametrize("path,dest_port",
+                             [("/system/v1/leader/mesos/foo/bar", 80),
+                              ("/mesos/master/state-summary", 5050),
+                              ])
```
+1

vespian reviewed (approved):
Just one more thing left - please re-use the endpoints we already have.

This change looks really good, I am OK with merging it. I would like us to come back to API quering at some point but till then - what we have here is definatelly better than what we have now in production.

🚂

mhrabovcin committed:
fix dcos-history-service upstream

jgehrcke commented on README.md:
```diff
@@ -32,6 +32,39 @@ The endpoint should only use relative links for links and referenced assets such
+[`proxy_pass`](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_pass) configuration directive that comes from a
+[variable](http://nginx.org/en/docs/http/ngx_http_rewrite_module.html#set).
+NGINX resolver respects DNS response TTL flags on each record so the resolve
+results are cached for period controlled by a upstream DNS server.
```
It's true that with the workaround we start respecting the TTL emitted by the DNS server, but we can also override it.

I propose a shorter version for this paragraph that
* skips the workaround explanation (instead refers to the well-written SO answer)
* documents the visible outbome

"The open source version of NGINX does not support periodic re-resolution of hostnames used in upstream definitions. In Admin Router, we use [a workaround](http://serverfault.com/a/593003/121951) to overcome this limitation, so that hostnames used in [proxy_pass](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_pass) configuration directives are guaranteed to be re-resolved periodically, either according to the TTL emitted by the DNS server or, if given, according to the `valid=` argument of the [resolver](http://nginx.org/en/docs/http/ngx_http_core_module.html#resolver) configuration."


Which brings us to the question: do we want to rely on the TTL emitted by Mesos DNS or do we want to make use of `valid=`?

mhrabovcin replied:
I've changed documentation.

In our case NGINX will be respecting TTL values from `mesos-dns` server that we control and its currently set to [`60` seconds](https://github.com/mesosphere/dcos-enterprise/blob/master/gen_extra/dcos-config.yaml#L148).

jgehrcke reviewed (changes requested):
I have left a few comments w.r.t. documentation / style. Looks great otherwise, we're ready to go ahead with this from my point of view.

mhrabovcin committed:
documentation fixes

mhrabovcin commented:
Integration tests are passing for both Open and EE version:

https://github.com/dcos/dcos/pull/1267
https://github.com/mesosphere/dcos-enterprise/pull/619

gpaul reviewed (approved):
:ship:

adam-mesos commented:
@mhrabovcin Do we need to update/port this to the dcos/dcos repo now that adminrouter lives there? Let's close out this PR and update https://jira.mesosphere.com/browse/DCOS-5809 when that's ready.

vespian commented:
Superseded by:
https://github.com/dcos/dcos/pull/1747
https://github.com/mesosphere/dcos-enterprise/pull/1205

Thanks @mhrabovcin !
